### PR TITLE
feat: add WFS building/building unit GRAR-275

### PR DESCRIPTION
**Pull request by ArneD**

### CODE REVIEW

File: src/BuildingRegistry.Projections.Wfs/Building/Building.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            b.Ignore(x => x.Version);
+
+            b.Property(p => p.GeometryMethod);
+
+            b.Property(p => p.Geometry)
+                .HasColumnType("sys.geometry");
+
+            b.Property(p => p.Status);
+
+            b.HasIndex(p => p.PersistentLocalId);
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/Building/Building.cs:67]: I thought we were going to put indexes on every field, no? Geometry?

File: src/BuildingRegistry.Projections.Wfs/Building/BuildingProjections.cs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+                await context.Buildings.AddAsync(new Building { BuildingId = message.Message.BuildingId });
+            });
+
+            When<Envelope<BuildingPersistentLocalIdWasAssigned>>(async (context, message, ct) =>
+            {
+                var building = await context.Buildings.FindAsync(message.Message.BuildingId, cancellationToken: ct);
+
+                building.Id = PersistentLocalIdHelper.CreateBuildingId(message.Message.PersistentLocalId);
+                building.PersistentLocalId = int.Parse(message.Message.PersistentLocalId.ToString());
+
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/Building/BuildingProjections.cs:41]: random newline

File: src/BuildingRegistry.Projections.Wfs/Building/BuildingProjections.cs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        private readonly WKBReader _wkbReader;
+
+        public BuildingProjections(WKBReader wkbReader)
+        {
+            _wkbReader = wkbReader;
+
+            When<Envelope<BuildingWasRegistered>>(async (context, message, ct) =>
+            {
+                await context.Buildings.AddAsync(new Building { BuildingId = message.Message.BuildingId });
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/Building/BuildingProjections.cs:32]: pass in `ct` to AddAsync
https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbset-1.addasync?view=efcore-3.1

File: src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+                foreach (var buildingUnitDetailItem in buildingUnits)
+                {
+                    if (message.Message.BuildingUnitIds.Contains(buildingUnitDetailItem.BuildingUnitId))
+                        unitsToRemove.Add(buildingUnitDetailItem);
+                }
+
+                var building = await context.BuildingUnitsBuildings.FindAsync(message.Message.BuildingId, cancellationToken: ct);
+                building.IsRemoved = true;
+
+                context.BuildingUnits.RemoveRange(unitsToRemove);
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs:55]: Hard delete? This implies they can never be reactivated, or no weird CRAB changes happen to them later

File: src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+            When<Envelope<BuildingStatusWasCorrectedToRemoved>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingStatusWasRemoved>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingWasCorrectedToPlanned>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingWasCorrectedToRealized>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingWasCorrectedToUnderConstruction>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingWasMeasuredByGrb>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingWasOutlined>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingWasPlanned>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingWasRealized>>(async (context, message, ct) => DoNothing());
+            When<Envelope<BuildingWasRegistered>>(async (context, message, ct) =>
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs:143]: Put this somewhere nicer, not right in the DoNothings :D

File: src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+
+            When<Envelope<CommonBuildingUnitWasAdded>>(async (context, message, ct) =>
+            {
+                await AddBuildingUnit(context, message.Message.BuildingId, message.Message.BuildingUnitId, BuildingUnitFunction.Common, message.Message.Provenance.Timestamp, ct);
+            });
+
+            When<Envelope<BuildingUnitWasRemoved>>(async (context, message, ct) =>
+            {
+                var buildingUnit = await context.BuildingUnits.FindAsync(message.Message.BuildingUnitId, cancellationToken: ct);
+                context.BuildingUnits.Remove(buildingUnit);
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs:197]: Hard delete here too?

File: src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+            });
+
+            When<Envelope<BuildingUnitWasAddedToRetiredBuilding>>(async (context, message, ct) =>
+            {
+                await AddBuildingUnit(context, message.Message.BuildingId, message.Message.BuildingUnitId, BuildingUnitFunction.Unknown, message.Message.Provenance.Timestamp, ct);
+                var addedUnit = await context.BuildingUnits.FindAsync(message.Message.BuildingUnitId, cancellationToken: ct);
+                var building = await context.BuildingUnitsBuildings.FindAsync(message.Message.BuildingId, cancellationToken: ct);
+
+                if (building.BuildingRetiredStatus == BuildingStatus.NotRealized)
+                    addedUnit.Status = GebouweenheidStatus.NietGerealiseerd.ToString();
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs:179]: In Building you put them nicely in constants, can you do this here too?

File: src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+                if (buildingUnit != null)
+                {
+                    buildingUnit.IsComplete = false;
+                    SetVersion(buildingUnit, message.Message.Provenance.Timestamp);
+                }
+            });
+
+            #endregion
+
+            #region BuildingUnitStatus
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/BuildingUnit/BuildingUnitProjections.cs:222]: In Building you put them nicely in constants, can you do this here too and use them in all the status handlers below?

File: src/BuildingRegistry.Projections.Wfs/Building/Building.cs (hunk 1 above)
Comment (CumpsD) [reply src/BuildingRegistry.Projections.Wfs/Building/Building.cs:67]: Ah they are in the migration, nvm

File: src/BuildingRegistry.Projections.Wfs/WfsContext.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+    public class WfsContext : RunnerDbContext<WfsContext>
+    {
+        public override string ProjectionStateSchema => Schema.Wfs;
+
+        public DbSet<Building.Building> Buildings { get; set; }
+        public DbSet<BuildingUnit.BuildingUnit> BuildingUnits { get; set; }
+        public DbSet<BuildingUnit.BuildingUnitBuildingItem> BuildingUnitsBuildings { get; set; }
+
+        //TODO: Add spatial indexes!
```
Comment (CumpsD) [inline src/BuildingRegistry.Projections.Wfs/WfsContext.cs:15]: this todo is done I guess? they are in your migration?

File: src/BuildingRegistry.Projections.Wfs/Building/Building.cs (hunk 1 above)
Comment (ArneD) [reply src/BuildingRegistry.Projections.Wfs/Building/Building.cs:67]: See 3th migration, can't add spatial indexes via fluent api.

**Pull request merged**